perf/sawtooth_perf/src/batch_map.rs

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::collections::HashMap;
+
+use sawtooth_sdk::messages::batch::BatchList;
+
+pub struct BatchMap {
+    // Batches that haven't been successfully submitted to
+    // a validator.
+    batches_by_id: HashMap<String, BatchList>,
+}
+
+impl BatchMap {
+    pub fn new() -> BatchMap {
+        BatchMap {
+            batches_by_id: HashMap::new(),
+        }
+    }
+
+    // Mark that batchlist associated with batch id has been submitted
+    // to a validator.
+    pub fn mark_submit_success(&mut self, batch_id: String) {
+        self.batches_by_id.remove(batch_id.as_str());
+    }
+
+    // Get a batchlist by id, to submit it to a validator.
+    pub fn get_batchlist_to_submit(&mut self, batch_id: String) -> Option<BatchList> {
+        self.batches_by_id
+            .clone()
+            .get(batch_id.as_str())
+            .map(|batchlist| batchlist.clone())
+    }
+
+    // Idempotent method for adding a BatchList
+    pub fn add(&mut self, batchlist: BatchList) {
+        let batch_id = match batchlist.batches.last() {
+            Some(batch) => Some(batch.header_signature.clone()),
+            None => None,
+        };
+        match batch_id {
+            Some(batch_id) => {
+                if !self.batches_by_id.contains_key(batch_id.as_str()) {
+                    self.batches_by_id.insert(batch_id.clone(), batchlist);
+                }
+            }
+            None => {}
```

Comment:
You can use `.map_or()` here instead, and you can chain it with the previous map: http://play.integer32.com/?gist=b31eda605c1d6d61c0d5aaf41ce048ae&version=stable